rct229/rulesets/ashrae9012019/ruleset_functions/compare_swh_dist_systems_and_components.py

```diff
@@ -0,0 +1,132 @@
+from rct229.rulesets.ashrae9012019.data_fns.extra_schema_fns import (
+    EXTRA_SCHEMA,
+    compare_context_pair,
+)
+from rct229.rulesets.ashrae9012019.ruleset_functions.get_swh_equipment_associated_with_each_swh_distriubtion_system import (
+    get_swh_equipment_associated_with_each_swh_distribution_system,
+)
+from rct229.utils.assertions import assert_
+from rct229.utils.utility_functions import (
+    find_exactly_one_pump,
+    find_exactly_one_service_water_heating_distribution_system,
+    find_exactly_one_service_water_heating_equipment,
+)
+
+
+def compare_swh_dist_systems_and_components(
+    rmd1: dict, rmd2: dict, compare_context_str: str, swh_distribution_id: str
+):
+    """This function compares all sub-components of a given SWH Distribution System ID between two models.
+    Parameters
+    ----------
+    rmd1: dict
+        RMD at RuleSetModelDescription level
+    rmd2: dict
+        RMD at RuleSetModelDescription level
+    compare_context_str: str
+    swh_distribution_id: str
+
+
+    Returns
+    -------
+    errors: list
+        An array of errors encountered in this function and in compare_context_pair. If errors is an empty array, then all elements that were expected to match do match
+
+    """
+
+    errors = []
+    rmd1_swh_system_and_equip_dict = (
+        get_swh_equipment_associated_with_each_swh_distribution_system(rmd1)
+    )
+    rmd2_swh_system_and_equip_dict = (
+        get_swh_equipment_associated_with_each_swh_distribution_system(rmd2)
+    )
+
+    # service_water_heating_distribution_systems
+    assert_(
+        rmd1_swh_system_and_equip_dict[swh_distribution_id],
+        f"SWH Distribution System {swh_distribution_id} not found in the `rmd1`.",
+    )
+    assert_(
+        rmd2_swh_system_and_equip_dict[swh_distribution_id],
+        f"SWH Distribution System {swh_distribution_id} not found in the `rmd2`.",
+    )
+    rmd1_swh_equipment_dict = rmd1_swh_system_and_equip_dict[swh_distribution_id]
+    rmd1_swh_distribution = find_exactly_one_service_water_heating_distribution_system(
+        rmd1, swh_distribution_id
+    )
+    rmd2_swh_equipment_dict = rmd2_swh_system_and_equip_dict[swh_distribution_id]
+    rmd2_swh_distribution = find_exactly_one_service_water_heating_distribution_system(
+        rmd2, swh_distribution_id
+    )
+
+    index_content_list, compare_context_list = (
+        (rmd1_swh_distribution, rmd2_swh_distribution)
+        if len(rmd1_swh_distribution) > len(rmd2_swh_distribution)
+        else (rmd2_swh_distribution, rmd1_swh_distribution)
+    )
+
+    compare_context_pair(
+        index_context=index_content_list,
+        compare_context=compare_context_list,
+        element_json_path="",
```

Comment:
Could @weilixu let me know what we should input for the ```element_json_path```  and ```search_key``` arguments?

Reply:
I think the search key is the same as the `compare_context_str`. `element_json_path` should starts with `"$"` 
In this case, it starts with `service_water_heating_distribution_system`, then maybe you can start with the jsonpath to the `service_water_heating_distribution_system`